### DISCUSSION TITLE
As a user, I would love it if the code would be able to automatically build the constrained expression, depending on what constraints I define.

### DISCUSSION THREAD

**Discussion by leakec** [leakec/tfc#19] [Q&A] [answered]
Is there a way to automatically generate the constrained expression given a set of constraints?

Original question from an email chain with Mario De Florio.

**Comment by leakec** [comment, marked as answer]
## TL;DR
Doing this with JAX works, but led to long compilation times and code that ran slower than if the constrained expression was written out by hand. However, you can definitely do this with symbolic solvers. The algorithm to do that is given [here](https://www.youtube.com/watch?v=uisOZVBHA2U&t=2082s&authuser=1) at ~26:13. There is a tool for that included in the TFC module, and there are examples in the repository that do it with Mathematica as well.

## Doing this with JAX
In terms of doing this directly with the code available in the TFC GitHub, I actually looked into this a while ago. Essentially, we need to loop over your constraints and then generate a Python function from them. This is totally possible and I even did it. The problem is that the automatic differentiation framework has issues here. Everything compiles and runs correctly, i.e., the solution is accurate, but the code runs slower. Something within their JIT compiler was having issues when the for loop was differentiated multiple times (plus the Jacobian if it's a non-linear problem). Also, the compilation itself was taking really long in this case. I may take another pass at this in the future since they improve things over time. However, there is a solution with symbolic solvers that I've found very useful.

## Doing this with a symbolic program
Even if doing this with TFC directly works well eventually, there are some perks to doing it symbolically: mainly analytical simplifications that would not be possible if we did it directly with automatic differentiation. For example, when automatic differentiation differentiates `x^2/2` I believe it should yield `2*x/2`, i.e., it doesn't cancel the 2 in the numerator and denominator. If we do at least the constrained expression analytically, we can get some analytical simplification at that level. 

The TFC module comes equipped with the `CeSolver` class that utilizes `sympy` to solve constrained expressions. There is a tutorial for using it included in the reference documentation [here](https://tfc-documentation.readthedocs.io/en/latest/notebooks/CeSolver.html).

I used to effectively have an analytical solver like this in Mathematica as well. In the TFC book problems (available in `examples` in the repository), Chapter 2 Problem 4 has an example of what this looks like. You'll need a copy of Mathematica or the free CDF viewer to view either the NB file or the CDF file (you can also browse back in the Git history and find when we had the PDFs). The section that looks like this is what I'm referring to (the formatting is horrible here since it is the old PDF, if you view this in Mathematica it is color coded and the symbols look much better):
![image](https://user-images.githubusercontent.com/46822212/216774555-2db37e5b-31e0-42c6-9d8b-ce72abb991df.png)

Essentially, you take your support functions and put them in `s1`. You take the constraint operators and put them in `S1`. Then, the rest is a simple algorithm that follows [this](https://www.youtube.com/watch?v=uisOZVBHA2U&t=2082s) (see around ~26:13 in the video for the correct slide). The PDF I'm referring to does this for a pretty complicated multivariate constrained expression, so it typically looks much simpler than what is shown there. I'm sorry I do not have a simpler working example that explains this better, but I don't have a copy of Mathematica to be able to produce one for you.
